packages/utils/src/cache/distributed-cache.ts

```diff
@@ -0,0 +1,307 @@
+import { Logger } from './helpers.js';
+import { Mutex } from 'async-mutex';
+import { CoreCache } from './core-cache.js';
+import { CacheMessenger } from './messenger.js';
+import type { CacheConfig } from './core-cache.js';
+import { TimeoutTaskRunner } from '../async.js';
+import { CACHE_PREFIX, MESSENGER_NO_MESSAGE_DELAY } from './constants.js';
+
+const logger = new Logger();
+logger.timeStart(window.name);
+
+/**
+ * Mutual exclusion, prevent getting requests state concurrently from multi tabs/windows
+ */
+const mutex = new Mutex();
+
+type StorageNames = {
+  requests: string
+  unloaded: string
+};
+
+type Requests = {
+  [key: string]: 'true'
+}
+
+/**
+ * Distribute cache across browsers windows, tabs, and iframes to reducing network requests
+ */
+export class DistributedCache extends CoreCache {
+
+  /**
+   * Requests cached from localStorage
+   */
+  protected requests: Requests = {};
+
+  /**
+   * List of promise needed be resolved by messaging
+   */
+  protected waiting = new Map<string, CallableFunction>();
+
+  /**
+   * Cache messenger for distribute cache
+   */
+  protected messenger: CacheMessenger;
+
+  /**
+   * Timer to check message is the last
+   */
+  protected noMessageTimeout = new TimeoutTaskRunner(MESSENGER_NO_MESSAGE_DELAY);
+
+  /**
+   * Names for manage all states temporary
+   */
+  protected storageNames: StorageNames;
+
+  /**
+   * Get resource request list, sync the new requests to cache
+   */
+  protected get requestsState (): Requests {
+    const state = localStorage.getItem(this.storageNames.requests);
+    if (!state) {
+      return {};
```

Comment:
What does `{}` value return represent ?

Reply:
Would it be better to use `null` to represent no request ?

Reply:
Yes, that is true, but I think it should always be the object because it makes coding easier/shorter that no need to check null every time before accessing the data inside.

Reply:
Personally I would choose null checks over a few lines of code.

Reply:
This code has been removed

Reply:
This code has been removed